src/lib/snarky/src/snark0.ml

```diff
@@ -915,159 +875,134 @@ module Make_basic (Backend : Backend_intf.S) = struct
-      let aux = Field.Vector.create () in
-      let get_value : Cvar.t -> Field.t =
-        let get_one v = Field.Vector.get aux (Backend.Var.index v - 1) in
-        Cvar.eval get_one
+        incr next_auxiliary ; v
       in
-      let store_field_elt x =
-        let v = Backend.Var.create !next_auxiliary in
-        incr next_auxiliary ;
-        Field.Vector.emplace_back aux x ;
-        v
+      let run_as_prover x s =
+        match (x, s) with
+        | Some x, Some s ->
+            let s', y = As_prover.run x get_value s in
+            (Some s', Some y)
+        | _, _ -> (None, None)
       in
-      let system = R1CS_constraint_system.create () in
-      R1CS_constraint_system.set_primary_input_size system 0 ;
+      ignore
+        (Option.map system ~f:(fun system ->
+             R1CS_constraint_system.set_primary_input_size system num_inputs )) ;
       let rec go : type a s.
-          string list -> (a, s) t -> Request.Handler.t -> s -> s * a =
+             string list
+          -> (a, s) t
+          -> Request.Handler.t
+          -> s option
+          -> s option * a =
        fun stack t handler s ->
         match t with
         | Pure x -> (s, x)
-        | With_constraint_system (f, k) -> f system ; go stack k handler s
+        | With_constraint_system (f, k) ->
+            ignore (Option.map ~f system) ;
+            go stack k handler s
         | With_label (lab, t, k) ->
             let s', y = go (lab :: stack) t handler s in
             go stack (k y) handler s'
         | As_prover (x, k) ->
-            let s', () = As_prover.run x get_value s in
+            let s', _ = run_as_prover (Some x) s in
             go stack k handler s'
         | Add_constraint (c, t) ->
-            if not (Constraint.eval c get_value) then
-              failwithf "Constraint unsatisfied:\n%s\n%s\n"
-                (Constraint.annotation c)
-                (Constraint.stack_to_string stack)
-                () ;
-            Constraint.add ~stack c system ;
+            ignore
+              (Option.map system ~f:(fun system ->
+                   (* NOTE: If s is None, we aren't evaluating and shouldn't consider constraints here. *)
+                   if Option.is_some s && not (Constraint.eval c get_value)
+                   then
+                     failwithf "Constraint unsatisfied:\n%s\n%s\n"
+                       (Constraint.annotation c)
+                       (Constraint.stack_to_string stack)
+                       () ;
+                   Constraint.add ~stack c system )) ;
             go stack t handler s
         | With_state (p, and_then, t_sub, k) ->
-            let s, s_sub = As_prover.run p get_value s in
+            let s, s_sub = run_as_prover (Some p) s in
             let s_sub, y = go stack t_sub handler s_sub in
-            let s, () = As_prover.run (and_then s_sub) get_value s in
+            let s, _ = run_as_prover (Option.map ~f:and_then s_sub) s in
```

Comment:
I think if this is a unit, better to just have the () pattern (it makes it more clear when reading that throwing this value away is fine).

Reply:
It's no longer a `unit`, it's a `unit option`. I've added a type signature instead, hopefully that makes it clearer.